test/xbgpu/test_engine.py

```diff
@@ -821,77 +818,91 @@ async def _send_data(
+                    # with the new values.
                     heap = await stream.get()
-                assert updated_items == {"frequency", "timestamp", "xeng_raw"}
-                # Ensure that the timestamp from the heap is what we expect.
-                assert (
-                    ig_recv["timestamp"].value % (timestamp_step * corrprod_output.heap_accumulation_threshold) == 0
-                ), "Output timestamp is not a multiple of timestamp_step * heap_accumulation_threshold."
 
-                assert (
-                    ig_recv["timestamp"].value
-                    == accumulation_index * timestamp_step * corrprod_output.heap_accumulation_threshold
-                ), (
-                    "Output timestamp is not correct. "
-                    f"Expected: "
-                    f"{hex(accumulation_index * timestamp_step * corrprod_output.heap_accumulation_threshold)}, "
-                    f"actual: {hex(ig_recv['timestamp'].value)}."
-                )
+                    while (updated_items := set(ig_recv.update(heap))) == set():
+                        # Test has gone on long enough that we've received another descriptor
+                        heap = await stream.get()
+                    assert updated_items == {"frequency", "timestamp", "xeng_raw"}
+                    # Ensure that the timestamp from the heap is what we expect.
+                    assert (
+                        ig_recv["timestamp"].value % (timestamp_step * corrprod_output.heap_accumulation_threshold) == 0
+                    ), "Output timestamp is not a multiple of timestamp_step * heap_accumulation_threshold."
+
+                    assert (
+                        ig_recv["timestamp"].value
+                        == accumulation_index * timestamp_step * corrprod_output.heap_accumulation_threshold
+                    ), (
+                        "Output timestamp is not correct. "
+                        f"Expected: "
+                        f"{hex(accumulation_index * timestamp_step * corrprod_output.heap_accumulation_threshold)}, "
+                        f"actual: {hex(ig_recv['timestamp'].value)}."
+                    )
 
-                assert ig_recv["frequency"].value == frequency, (
-                    "Output channel offset not correct. "
-                    f"Expected: {frequency}, "
-                    f"actual: {ig_recv['frequency'].value}."
-                )
+                    assert ig_recv["frequency"].value == frequency, (
+                        "Output channel offset not correct. "
+                        f"Expected: {frequency}, "
+                        f"actual: {ig_recv['frequency'].value}."
+                    )
 
-                corrprod_results[i][j] = ig_recv["xeng_raw"].value
+                    corrprod_results[corrprod_output.name][j] = ig_recv["xeng_raw"].value
+            except spead2.Stopped:
+                pass
```

Comment:
I'm also not wild about it. My main worry is that if something causes a Stopped exception in one of the other tests, you're going to have zeros in the results and you'll get a data mismatch, which will be misleading and make the failure harder to debug.

My inclination would be to pass extra information into send_data to indicate when it should expect to run out of data. It can then directly verify that this is actually when data stops flowing: after receiving all the expected data, it could try to receive another heap inside a `pytest.raises` context manager (which might still need to handle descriptors).

It might also be nice to then size the results array to the amount of data actually expected, instead of returning arrays with trailing zeros. That's actually what I thought you were going to do when I saw you'd replaced a big array with a dictionary of arrays. I realise this could have knock-on implications for all the verification functions though, so if it's going to be big pain then don't bother.

Reply:
> I realise this could have knock-on implications for all the verification functions though, so if it's going to be big pain then don't bother.

Your proposed suggestion is exactly what I intended on doing. Stopped short of that because I didn't want my changes to be too wholesale. Still, I see it's tending that way anyway even with just a "expect data to stop here" parameter. Might as well go all in.